### ISSUE TITLE
Enable multiselect with ctrl + shift

### ISSUE THREAD

**Issue by frankbolviken** [DevExpress/devextreme-reactive#532]
It would be of great use if you could multi select rows in the grid with holding in shift for instance and every row in the middle of the first you have selected and the next selection will also be selected.

Is a feature like this planned implemented?

**Comment by kvet** [comment]
Hi,
 
Thank you for providing us with a good suggestion. We will consider adding this feature in the future. In the meantime, you can try implementing it on your side. If you need any assistance with implementation details, let us know.

**Comment by frankbolviken** [comment]
Hi!

You guys are probably the most quick to reply on issues posted. Good job!
I was thinking something in the lines of checking if a key was pressed down while doing the selection.
Do you have any quick examples? If not it's probably not that hard to implement i would believe :)

Keep up the good work!

**Comment by kvet** [comment]
Here is an example implementation of this behavior: https://stackblitz.com/edit/react-nhnua1?file=index.js
 
I've also found a small issue that prevents second execution of an action in one React rendering transaction. It will be processed in a separate issue: https://github.com/DevExpress/devextreme-reactive/issues/549

**Comment by DanHarman** [comment]
Hi

Any chance of you guys supporting this fully in the grid itself? I've had a look at the example implementation and it is quite lacking in correct functionality which I think would be a lot of effort to implement for all your users. The issues I spotted immediately:

- ctrl-click does not clear a selection.
- when shift-clicking to select a range. If you do not lift up the shift key and and select another row, it should select that range instead and forget the range from that shift select operation. e.g. I have row 1 selected, I shift click row 3, I now have rows 1-3 selected. If I now click row 2, without lifting the shift key, the range should actually be rows 1-2. However in the example it is still rows 1-3.

Thanks,

Dan

**Comment by kvet** [comment]
Hi @DanHarman,

We are currently collecting information on this feature request and greatly appreciate your feedback. I am not sure how soon we will implement it, so at this time I suggest you use the following workaround: https://stackblitz.com/edit/react-4bhpkg?file=index.js. It contains the described functionality.

**Comment by DanHarman** [comment]
@kvet Hi thanks for that workaround, it behaves much more consistently with expectations!

**Comment by DanHarman** [comment]
@kvet I have run into a small problem. The selection is super fast which I like, but I was using  `<SelectionState selection={this.state.selection} onSelectionChange={this.handleSelectionChange} />`.

Is there a way to access the selection state and detect changes with your workaround? When I add the `selection` and `onSelectionChange` properties the multi select doesn't work.

**Comment by kvet** [comment]
@DanHarman, I'm not sure that I understand the problem. Here is the working example with the controlled mode: https://stackblitz.com/edit/react-p3fueb?file=index.js

**Comment by DanHarman** [comment]
@kvet This is strange, in my project I have a problem when there is a plugin for sorting or filtering incorporated. It no longer selects a range when holding down shift. I'm trying to isolate what the issue is as I tried adding them into your demo and they seem to work there. 

Next step is to put your demo into my app and see if it still works...

**Comment by DanHarman** [comment]
@kvet Hi, took a while to work out the problem, but its because I was regenerating the columns list in the render method (due to it needing some props which were functions used in getCellValue for a few columns). I'm not sure why this would cause a problem, but its not terribly efficient so I moved the columns to the constructor and all seems well.

**Comment by DanHarman** [comment]
@kvet actually maybe you can help me. I store lookup data in my redux store, and this data is loaded in parallel with the grid data. Previously when the prop containing this ref data was updated, the grid would rerender because I was creating the columns collection in the render method (and passing in the updated refdata to the function in getCellData).

Now that I create the columns in the constructor this is broken. I could use a lifecycle method to detect property changes and regenerate the columns collection, but it seems like your plugin doesn't like a changing columns collection so I'm not sure how to make this work?

**Comment by DanHarman** [comment]
@kvet Hi. Thanks again for this. Users are very happy. However is there a way I can wire in selectall functionality? I don't want the column with the checkboxes etc from the standard `tableselection` plugin, but I need to be able to wire a button that can do this (as well as capture ctrl-a). So maybe taking a ref to the `tableselectionadvanced` and exposing `toggleSelectAll` on that?

I did look at the tableselection plugin to see if I could copy across enough code to make it work, but not having much luck as I suspect it requires a pretty deep understanding of your architecture.

**Comment by kvet** [comment]
Hi,

We are working on the documentation describing how to use plugin export. They can be linked to your custom component with the TemplateConnector component. It can be treated like the redux' connect component. The TemplateConnector component will not work outside the Grid component.

I've also prepared an example demonstrating the use of TemplateConnector for adding the Select All button: https://stackblitz.com/edit/react-nyprct?file=index.js

**Comment by DanHarman** [comment]
@kvet Thanks for the quick turn around. I've run into a small problem and hoping you can help?

How can I get the plugin to work without the `<Toolbar />` plugin, as when I remove it, the `<SelectAllToolbarItem />` is not rendered.

I forked your code and got to get to this:

https://stackblitz.com/edit/react-vjqk7x?file=index.js

It has the empty toolbar rendering which I do not want as the 'select all' button needs to be external to the grid (hence the use of ref).

Also, what is the purpose of the ` <TemplatePlaceholder />` in `SelectAllToolbarItem`?

Thanks for the help,

Dan

**Comment by DanHarman** [comment]
Ok think I've worked it out -  looks like I just need to change the template name to table for the selectalltoolbaritem and it works.

**Comment by simoneconti** [comment]
I'm trying to implement the TableSelectionAdvanced plugins in typescript but I'm having hard time with that. Can you please give a solution for typescript users?
Thank you,

**Comment by SergeyAlexeev** [comment]
Hi,

Would you please clarify what exact problems you are experiencing?

**Comment by simoneconti** [comment]
> Hi,
> 
> Would you please clarify what exact problems you are experiencing?

I'm having trouble for example with the predicate of the Template component.

You can see it there:
https://stackblitz.com/edit/react-ts-grid-plugin-selection-advanced

**Comment by SergeyAlexeev** [comment]
Thank you for the additional information.

The first issue I found is that you are using different versions of DevExtreme Reactive packages: `1.8.0-beta.1` and `1.8.0`. I suggest you update all packages to version `1.8.0`.

The second issue is related to the Template predicate. Your code always returns false because you are comparing the Symbol and String types:

```
predicate={({ tableRow }) => {
  // typeof tableRow.type -> "symbol"
  return tableRow.type === 'data';
}}
```

Use the following code to fix the problem:

```
predicate={({ tableRow }) => {
  return tableRow.type === Table.ROW_TYPE;
}}
```

Please refer to this [article](https://devexpress.github.io/devextreme-reactive/react/grid/docs/reference/table#static-fields) for more details.
See the updated [sample](https://stackblitz.com/edit/react-ts-grid-plugin-selection-advanced-hyir82?file=index.tsx), as well.

**Comment by simoneconti** [comment]
> Thank you for the additional information.
> 
> The first issue I found is that you are using different versions of DevExtreme Reactive packages: `1.8.0-beta.1` and `1.8.0`. I suggest you update all packages to version `1.8.0`.
> 
> The second issue is related to the Template predicate. Your code always returns false because you are comparing the Symbol and String types:
> 
> ```
> predicate={({ tableRow }) => {
>   // typeof tableRow.type -> "symbol"
>   return tableRow.type === 'data';
> }}
> ```
> Use the following code to fix the problem:
> 
> ```
> predicate={({ tableRow }) => {
>   return tableRow.type === Table.ROW_TYPE;
> }}
> ```
> Please refer to this [article](https://devexpress.github.io/devextreme-reactive/react/grid/docs/reference/table#static-fields) for more details.
> See the updated [sample](https://stackblitz.com/edit/react-ts-grid-plugin-selection-advanced-hyir82?file=index.tsx), as well.

Thank you, but I'm using some tslint rules that are giving me some errors I don't know how to fix.
For example when declaring:
```
rows;
lastSelectedId;
lastMultipleSelection;
```
I get a "Member implicitly has an 'any' type."

In:
`handleClick({ row, getRowId, selection, rows, toggleSelection }, e)`
I get "expected parameter: '{ row, getRowId, selection, rows, toggleSelection }' to have a typedef"

In
`className={selection.indexOf(getRowId(params.tableRow.row)) > -1 ? 'active' : ''}`
I get "Property 'tableRow' does not exist on type 'object'."
 and so on.

I would like to not disable these rules, so I have to define everything.

Thank you.

**Comment by katzi0** [comment]
Hi,
the example https://stackblitz.com/edit/react-whavyh?file=index.html
doesn't work with the new versions. Any clues how to make this work again?

**Comment by SergeyAlexeev** [comment]
Hi,

As I [mentioned](https://github.com/DevExpress/devextreme-reactive/issues/532#issuecomment-435035028) above:

> The second issue is related to the Template predicate. Your code always returns false because you are comparing the Symbol and String types.

[Here](https://stackblitz.com/edit/react-xcfz1i?file=index.js) is the updated sample.

**Comment by churkin** [comment]
Duplicate https://github.com/DevExpress/devextreme-reactive/issues/1966
